harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,412 @@
+        }
+
+
+class _CPU(_MetricGroupCollector):
+    group = "cpu"
+
+    def collect(self) -> Dict[str, Any]:
+        cpu_util = psutil.cpu_percent()
+        return {
+            "cpu_util_simple": cpu_util,
+        }
+
+
+class _GPU(_MetricGroupCollector):
+    """
+    Collects GPU metrics for all GPU devices on the machine.
+
+    Returns `gpu_util` and `gpu_free_memory` metrics labeled by the device's UUID. Example:
+
+    .. code::
+        {
+            "GPU-UUID-A": {
+                "gpu_util": 0.12,
+                "gpu_free_memory": 5.662642176,
+            },
+            "GPU-UUID-B": {
+                "gpu_util": 0.23,
+                "gpu_free_memory": 6.552642176,
+            },
+        }
+    """
+
+    group = "gpu"
+
+    def __init__(self):
+        super().__init__()
+
+        self._pynvml_device_handles: Dict[str, Any] = {}
+
+        if pynvml:
+            self._init_pynvml()
+        else:
+            logging.warning(f"pynvml module not found. GPU metrics will not be collected.")
+
+    def _init_pynvml(self) -> None:
+        try:
+            pynvml.nvmlInit()
+            num_gpus = pynvml.nvmlDeviceGetCount()
+            for i in range(num_gpus):
+                handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                uuid = pynvml.nvmlDeviceGetUUID(handle)
+                pynvml.nvmlDeviceGetMemoryInfo(handle)
+                pynvml.nvmlDeviceGetUtilizationRates(handle)
+                self._pynvml_device_handles[uuid] = handle
+        except pynvml.NVMLError as ne:
+            logging.info(f"Error accessing NVML {ne}. GPU metrics will not be collected.")
+        except Exception as e:
+            logging.error(f"Error initializing pynvml: {e}")
+
+    def collect(self) -> Dict[str, Any]:
+        metrics = {}
+
+        for uuid, handle in self._pynvml_device_handles.items():
+            gpu_util = pynvml.nvmlDeviceGetUtilizationRates(handle).gpu
+            free_memory = pynvml.nvmlDeviceGetMemoryInfo(handle).free
+            metrics.update(
+                {
+                    uuid: {
+                        "gpu_util": gpu_util,
+                        "gpu_free_memory": free_memory,
+                    }
+                }
+            )
+        return metrics
+
+
+class _Metric:
+    def __init__(
+        self,
+        group: str,
+        metrics: Dict[str, Any],
+        timestamp: datetime,
+    ):
+        self.group = group
+        self.metrics = metrics
+        self.timestamp = timestamp
+
+
+class _Collector(threading.Thread):
+    def __init__(
+        self,
+        metrics_queue: queue.Queue,
+        collection_interval: int = 1,
+        max_collection_s: int = 60 * 30,
+    ):
+        self._collection_interval = collection_interval
+        self._should_exit = False
+        self._last_flush_ts = None
+        self._metrics_queue = metrics_queue
+        self._max_collection_s = max_collection_s
+
+        self._metric_collectors = [
```

Comment:
I'm finding myself wanting simpler `init`s in places in this file and more dependency injection. But I haven't worked out a good design or if it's a reasonable thing to expect. Chat?